### PR TITLE
Add in qpy the support for Discriminator and Kernel

### PR DESCRIPTION
**Pull request by junnaka51**
### Summary
This PR intends to solve #10050 by adding in `qpy` the support for `Discriminator` and `Kernel`. 
This enables users to serialize `Acquire` instructions with `Discriminator` and `Kernel` using `qpy`.

### Details and comments

The following are from #10050 and show that the error is resolved:
```py
In [1]: from io import BytesIO
   ...: 
   ...: import numpy as np
   ...: 
   ...: from qiskit import QuantumCircuit, pulse
   ...: from qiskit.qpy import dump, load
   ...: 
   ...: 
   ...: qc = QuantumCircuit(1, 1)
   ...: qc.measure(0, 0)
   ...: 
   ...: discriminator = pulse.Discriminator("hw_centroid")
   ...: 
   ...: with pulse.builder.build() as sched:
   ...:     pulse.builder.acquire(10, 0, pulse.MemorySlot(0), discriminator=discriminator)
   ...: 
   ...: qc.add_calibration("measure", (0,), sched)
   ...: 
   ...: file_ = BytesIO()
   ...: dump(qc, file_)
   ...: file_.seek(0)
   ...: new_qc = load(file_)[0]
   ...: print(qc == new_qc)
True

In [2]: 
```
```py
In [1]: from io import BytesIO
   ...: 
   ...: import numpy as np
   ...: 
   ...: from qiskit import QuantumCircuit, pulse
   ...: from qiskit.qpy import dump, load
   ...: 
   ...: 
   ...: qc = QuantumCircuit(1, 1)
   ...: qc.measure(0, 0)
   ...: 
   ...: kernel = pulse.Kernel("hw_qmfk", kernel={"real": np.zeros(10), "imag": np.zeros(10)}, bias=[0, 0])
   ...: 
   ...: with pulse.builder.build() as sched:
   ...:     pulse.builder.acquire(10, 0, pulse.MemorySlot(0), kernel=kernel)
   ...: 
   ...: qc.add_calibration("measure", (0,), sched)
   ...: 
   ...: file_ = BytesIO()
   ...: dump(qc, file_)
   ...: file_.seek(0)
   ...: new_qc = load(file_)[0]
   ...: print(qc == new_qc)
True

In [2]: 
```

### CODE REVIEW

File: qiskit/qpy/type_keys.py
Comment (wshanks) [inline qiskit/qpy/type_keys.py:332]: We should remove the preceding comment.

File: qiskit/qpy/type_keys.py
Comment (wshanks) [inline qiskit/qpy/type_keys.py:67]: Minor but I would put this alphabetically between the `qiskit.pulse.channels` and `qiskit.pulse.instructions` imports.

File: qiskit/pulse/configuration.py
Comment (wshanks) [inline qiskit/pulse/configuration.py:27]: We need to propagate up the case where they are not equal:

```suggestion
                if not _assert_nested_dict_equal(a[key], b[key]):
                    return False
```

File: qiskit/pulse/configuration.py
Comment (wshanks) [inline qiskit/pulse/configuration.py:24]: We need to ensure b does not have extra keys. Here is one way

```suggestion
def _assert_nested_dict_equal(a, b):
    if len(a) != len(b):
        return False
    for key in a:
```

File: qiskit/pulse/configuration.py
Comment (wshanks) [inline qiskit/pulse/configuration.py:66]: Was there a need for the hash? I didn't notice the kernel being used in a set or dictionary.

**Review by wshanks — COMMENTED** [review]
This generally looks good to me. The tests demonstrate the behavior we want. Based on my comments on the nested dict equality function, we might want some tests of Kernel/Discriminator equality.

A qpy expert should review the qpy changes (@jakelishman @mtreinish @nkanazawa1989). A qpy the LIST handling concerns me a bit -- it only works for json-serializable lists. Also, qpy tries to be compact, so falling back to json feels like it is out of character. I wonder if the sequence type could be used? One issue is that the kernel/discriminator parameters are open-ended in schema (I guess whatever the qobj json serializer supports).

File: qiskit/qpy/common.py
Comment (TsafrirA) [inline qiskit/qpy/common.py:23]: Should QPY version be bumped?

File: qiskit/qpy/binary_io/schedules.py
Comment (TsafrirA) [inline qiskit/qpy/binary_io/schedules.py:64]: Unless I am missing a technical reason to have these two objects handled in the same function, I think I would have gone for two separate functions.

**Review by TsafrirA — COMMENTED** [review]
This is still WIP, so I am guessing the documentation and release notes will follow later, but I do want to raise the question if QPY version should be bumped.

Additionally, the [compatibility test](https://github.com/Qiskit/qiskit-terra/blob/main/test/qpy_compat/test_qpy.py) should be updated.

File: qiskit/pulse/configuration.py
Comment (junnaka51) [reply qiskit/pulse/configuration.py:69]: If `Kernel` is unhashable, the following test fails:
https://github.com/Qiskit/qiskit-terra/blob/b73115f80bc575b5dbc8b98ad003cd3b4ff9c00d/test/python/pulse/test_instructions.py#L68-L101

Comment (wshanks) [reply qiskit/pulse/configuration.py:69]: Hmm, I am not sure what the best approach is here.

The current behavior is that the kernel and discriminator use Python's default hashing and equality behavior where every instance gets a unique hash and the only equality is that an instance is equal to itself. So they are hashable but not in a particularly useful way. Overriding the default `__eq__` method unsets the default `__hash__` method. I wonder if other behavior relies on pulse instructions being hashable? I have not tried to check.

The kernel and discriminator classes are not well-specified -- basically just json serializable structures (extended to support numpy arrays) -- so nested dicts and lists of float, int, str, and ndarrays. That worked fine for qobj but does not fit as well for qpy. They don't need to be mutable so hashing the data should be okay. I think a more proper hash would be to replace lists with tuples and dicts with tuples of key-value tuples recursively, but that is a bit involved, especially if we don't really need hashing.

File: qiskit/qpy/type_keys.py
Comment (junnaka51) [reply qiskit/qpy/type_keys.py]: Done! :) 👉 https://github.com/Qiskit/qiskit-terra/pull/10327/commits/ed7515dd5aff98120824998f3476d1a271c7dbbd

File: qiskit/qpy/type_keys.py
Comment (junnaka51) [reply qiskit/qpy/type_keys.py:328]: Done! :) 👉 https://github.com/Qiskit/qiskit-terra/pull/10327/commits/ed7515dd5aff98120824998f3476d1a271c7dbbd

**Comment by junnaka51** [comment]
@wshanks 

> Based on my comments on the nested dict equality function, we might want some tests of Kernel/Discriminator equality.

Thank you so much for your comments! I added those tests :) 👉 https://github.com/Qiskit/qiskit-terra/pull/10327/commits/5cf5d3a3a1b91e551dda8ff12f0cb408e6bc4f8f

File: qiskit/pulse/configuration.py
Comment (junnaka51) [reply qiskit/pulse/configuration.py]: Thank you very much for pointing out my error! 🙇

**Comment by junnaka51** [comment]
@TsafrirA 

> Additionally, the [compatibility test](https://github.com/Qiskit/qiskit-terra/blob/main/test/qpy_compat/test_qpy.py) should be updated.

Thank you very much for pointing it out! I did not know it.
I added a test here 👉 https://github.com/Qiskit/qiskit-terra/pull/10327/commits/36f628cc4f94fbf8501f9e24fc87946de12d8a5a
But I am not familiar with this kind of compatibility tests, so any suggestion for improving this would be very much appreciated 🙇

File: qiskit/pulse/configuration.py
Comment (junnaka51) [reply qiskit/pulse/configuration.py]: Thank you!

File: qiskit/qpy/common.py
Comment (junnaka51) [reply qiskit/qpy/common.py:23]: I was not sure. I would like to hear from a QPY expert.

File: qiskit/qpy/type_keys.py
Comment (wshanks) [inline qiskit/qpy/type_keys.py:144]: I have not dug deeply into the internals of qpy. Why do we need these `LIST` and `DICT` types? Why can we not just use `SEQUENCE` and `MAPPING`? (I see that `MAPPING` did not support nesting and  you extended it with `DICT` but could it not be extended with `DICT`?).

**Review by wshanks — COMMENTED** [review]
This is looking pretty good to me.

I do not like the hashing based on string representation and there are some questions about the qpy details, but I think this is looking good otherwise.

File: qiskit/pulse/configuration.py
Comment (junnaka51) [reply qiskit/pulse/configuration.py:69]: > I wonder if other behavior relies on pulse instructions being hashable? I have not tried to check.

`SymbolicPulse` is not hashable:
https://github.com/Qiskit/qiskit-terra/blob/e9f8b7c50968501e019d0cb426676ac606eb5a10/qiskit/pulse/library/symbolic_pulses.py#L570

So, not all of instructions are actually hashable.

Comment (junnaka51) [reply qiskit/pulse/configuration.py:89]: >  I think a more proper hash would be to replace lists with tuples and dicts with tuples of key-value tuples recursively, but that is a bit involved, especially if we don't really need hashing.

Thank you so much for your suggestion!
Although I am still not sure whether `Kernel` and `Discriminator` should remain hashable (because the other operand `SymbolicPulse` is unhashable and as a result not all of instructions are hashable), as long as we cannot eliminate the possibility that their hashing is used somewhere, I suppose that it is safer to keep them hashable. 
I updated the hash methods following your suggestion :) 👉  https://github.com/Qiskit/qiskit-terra/pull/10327/commits/0afd18dde0e3041eb3ae7e40492ea55d11fdb89d

File: qiskit/qpy/common.py
Comment (nkanazawa1989) [inline qiskit/qpy/common.py:175]: Instead of doing this, you can generate `datum_bytes` with this trick by recursively calling `write_mapping` with `container` in the code below.
https://github.com/Qiskit/qiskit-terra/blob/e9f8b7c50968501e019d0cb426676ac606eb5a10/qiskit/qpy/common.py#L196-L198

File: qiskit/qpy/type_keys.py
Comment (nkanazawa1989) [reply qiskit/qpy/type_keys.py:144]: I agree. @junnaka51 is there any reason not to use existing types?

File: test/python/pulse/test_experiment_configurations.py
Comment (nkanazawa1989) [inline test/python/pulse/test_experiment_configurations.py:97]: 💯

File: test/python/pulse/test_experiment_configurations.py
Comment (nkanazawa1989) [inline test/python/pulse/test_experiment_configurations.py:178]: 💯

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @junnaka51 for the continuous contribution. The PR looks great. Let me add some comments.

File: qiskit/pulse/configuration.py
Comment (wshanks) [reply qiskit/pulse/configuration.py:23]: How would you do that without making the provider a (circular) dependency of terra?

I think this function is okay. Maybe a comment could be added that the only reason it is needed is because `ndarray` overrides `__eq__` in an unhelpful way? The other things the JSON encoder handles are `complex` and `ParameterExpression` which work okay with `__eq__`.

File: qiskit/pulse/configuration.py
Comment (wshanks) [inline qiskit/pulse/configuration.py:42]: This looks good to me, but following @nkanazawa1989's comment I think just removing hashing might be better since we can't think of a good use case and less code to maintain is always better.

File: qiskit/qpy/common.py
Comment (wshanks) [reply qiskit/qpy/common.py:23]: This guidance was a little surprising to me at first since we are adding a new data type, but I think it makes sense. Here is how I think about it that makes sense for me:

qpy attempts to provide forwards compatibility for loading previously stored circuits (I think there is still freedom for terra to completely remove something and then it is okay not to load that any more). Since the way data is stored for a specific object might need to change, qpy versions its files and can use the version number to know to load an object stored with the old format. Here we are adding new types to be stored, so the change has no impact on loading previously stored data. The part that was confusing to me is that we could store a kernel in terra 0.25 with qpy version 7 and terra 0.24 would also have qpy version 7 but not be able to load the kernel. However, the terra version is also stored in the qpy data and a separate warning is generated when trying to load qpy data with an earlier version of terra than the one used to generate the data, so I think that covers that case.

File: qiskit/pulse/configuration.py
Comment (nkanazawa1989) [reply qiskit/pulse/configuration.py:23]: Yes, basically I agree with you. Unfortunately likely Qiskit doesn't have control of Kernel/Discriminator format, so we should continuously maintain this function to sync with the provider. I was thinking it was bit easier to delegate everything to the provider, but from the dependency viewpoint this approach is bit tough. 

I think at some point we should define subclass and use something like [pydantic](https://docs.pydantic.dev/latest/) rather than this free-form dictionary. Of course not in this PR :)

File: qiskit/qpy/type_keys.py
Comment (junnaka51) [reply qiskit/qpy/type_keys.py:144]: I cannot find `SEQUENCE` and `MAPPING` as types in https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qpy/type_keys.py . 

If I understand correctly, we need an identifier of `LIST` (or we can call it `SEQUENCE`) and that of `DICT` (or `Mapping`) for reading them out from a binary. 
I now realize that we can use the existing functions such as
https://github.com/Qiskit/qiskit-terra/blob/e9f8b7c50968501e019d0cb426676ac606eb5a10/qiskit/qpy/common.py#L203
 instead of upgrading `qpy.binary_io.value.dumps_value()` and `qpy.binary_io.value.loads_value()`. But I think we still need the identifiers. For example, when we want to deal with `LIST` as a value in a `DICT`, then we may update `qpy.common.write_mapping()` as
```diff
def write_mapping(file_obj, mapping, serializer, **kwargs):
    """Write a mapping in the file like object.
    """
    num_elements = len(mapping)

    file_obj.write(struct.pack(formats.SEQUENCE_PACK, num_elements))
    for key, datum in mapping.items():
        key_bytes = key.encode(ENCODE)

+        if isinstance(datum, list):
+            type_key, datum_bytes = b"l", sequence_to_binary(datum, serializer, **kwargs)
+        else:
+            type_key, datum_bytes = serializer(datum, **kwargs)

        item_header = struct.pack(formats.MAP_ITEM_PACK, len(key_bytes), type_key, len(datum_bytes))
        file_obj.write(item_header)
        file_obj.write(key_bytes)
        file_obj.write(datum_bytes)
```
and `qpy.common.read_mapping()` as
```diff
def read_mapping(file_obj, deserializer, **kwargs):
    """Read a mapping from the file like object.
    """
    mapping = {}

    data = formats.SEQUENCE._make(
        struct.unpack(formats.SEQUENCE_PACK, file_obj.read(formats.SEQUENCE_SIZE))
    )
    for _ in range(data.num_elements):
        map_header = formats.MAP_ITEM._make(
            struct.unpack(formats.MAP_ITEM_PACK, file_obj.read(formats.MAP_ITEM_SIZE))
        )
        key = file_obj.read(map_header.key_size).decode(ENCODE)

+        if map_header.type == b"l":
+            datum = common.read_sequence(file_obj, deserializer, **kwargs)
+        else:
+             datum = deserializer(map_header.type, file_obj.read(map_header.size), **kwargs)
-        datum = deserializer(map_header.type, file_obj.read(map_header.size), **kwargs)
        mapping[key] = datum

    return mapping
```
, in both of which `b"l"` is used as the identifier of `LIST` (or `SEQUENCE` if we want). 
Of course this `b"l"` must not be defined in `qpy/type_keys.py`. But I thought `qpy/type_keys.py` summarizes all the types, so I put it there.

Comment (nkanazawa1989) [reply qiskit/qpy/type_keys.py:144]: You're absolutely right. I think nested dictionary is not a `Value` type, and indeed I didn't assume this data structure when I wrote this function. Instead of calling `value.dumps_value`, you can write a custom serializer function for discriminator/kernel config, which would look like

```python
def my_custom_serializer(obj, version):
    if isinstance(obj, dictionary):
	 with io.BytesIO() as container: 
            for key, value in obj.items()
                # recursively generate binary data
            binary_data = container.getvalue() 
        return b"D", binary_data
    return dumps_value(obj, version)
```

what do you think?

Comment (junnaka51) [reply qiskit/qpy/type_keys.py:144]: @nkanazawa1989  Thank you so much for your suggestion 🙇 
Please give me a little more time. It has been taking time to understand it.

Comment (wshanks) [reply qiskit/qpy/type_keys.py:144]: Thanks, @junnaka51! By the way, this PR is currently labeled with the 0.25.0 milestone which should be released in about 10 days. There is no pressure to meet the milestone though. We can bump it back to the next milestone if needed. Leaving the milestone on the PR for now means that we can still merge it if it is ready (a newly opened PR would not generally by eligible for 0.25.0 at this point).

Comment (junnaka51) [reply qiskit/qpy/type_keys.py]: @wshanks Thank you for the info! :)

Comment (wshanks) [reply qiskit/qpy/type_keys.py]: In any case, I think with your last push this PR might be good now? The one thing I wonder about is the lack of `type_keys` for `b"D"` and `b"l"` -- it feels like names would be clearer than plain strings, but I defer to @nkanazawa1989

Comment (junnaka51) [reply qiskit/qpy/type_keys.py]: No, not yet. Although it works, I suppose how to deal with a nested dictionary is not yet well written.
I am trying to grasp @nkanazawa1989 `s comments.

Comment (junnaka51) [reply qiskit/qpy/type_keys.py]: @nkanazawa1989 Thank you again for your advice and suggestion 🙇 
I agree that your approach is more sophisticated. I added a new serializer and deserializer comparable in abstraction to `value.dumps_value` and `value.loads_value`, respectively, and put `write_mapping` and `read_mapping` back to their original 👉 https://github.com/Qiskit/qiskit-terra/pull/10327/commits/cc878e657bffb5a671c07637de86322dbf537eae

**Review by wshanks — APPROVED** [review]
This all looks totally functional and well structured to me. I don't quite follow why the nested sequence and mapping serialization/deserialization needs to be kept out of the main loads_value/dumps_value functions, but it matches what Naoki suggested.

@nkanazawa1989 Do you want to give final approval and get this merged in time for the next release?

**Review by mtreinish — DISMISSED** [review]
From the QPY side this LGTM, we might want some docs on how `KERNEL` and `DISCRIMINATOR` are serialized. But it's not critical and we can do it post merge. The one thing to point out is that this will likely merge conflict with #10392 but the rebase shouldn't be too bad.

**Comment by junnaka51** [comment]
@wshanks Thank you so much for resolving the conflict! 🙇

**Review by mtreinish — APPROVED** [review]

**Comment by junnaka51** [comment]
@wshanks @nkanazawa1989 @TsafrirA @mtreinish Thank you so much for review! 🙇‍♂️

**Pull request merged**
